pdc/apps/repository/views.py

```diff
@@ -356,3 +357,87 @@ def list(self, request, *args, **kwargs):
         %(SERIALIZER)s
         """
         return super(ServiceViewSet, self).list(request, *args, **kwargs)
+
+
+class PushTargetViewSet(PDCModelViewSet):
+    """
+    Push targets for products, product versions, releases and release variants.
+    """
+
+    queryset = models.PushTarget.objects.all()
+    serializer_class = serializers.PushTargetSerializer
+    filter_class = filters.PushTargetFilter
+    permission_classes = (APIPermission,)
+
+    def create(self, request, *args, **kwargs):
+        """
+        __Method__: POST
+
+        __URL__: $LINK:pushtarget-list$
+
+        __Data__:
+
+        %(WRITABLE_SERIALIZER)s
+
+        __Response__:
+
+        %(SERIALIZER)s
+        """
+
+        return super(PushTargetViewSet, self).create(request, *args, **kwargs)
+
+    def retrieve(self, request, *args, **kwargs):
+        """
+        __Method__: GET
+
+        __URL__: $LINK:pushtarget-detail:instance_pk$
+
+        __Response__:
+
+        %(SERIALIZER)s
+        """
+        return super(PushTargetViewSet, self).retrieve(request, *args, **kwargs)
+
+    def list(self, *args, **kwargs):
+        """
+        __Method__: GET
+
+        __URL__: $LINK:pushtarget-list$
+
+        __Query params__:
+
+        %(FILTERS)s
+
+        __Response__: a paged list of following objects
+
+        %(SERIALIZER)s
+        """
+        return super(PushTargetViewSet, self).list(*args, **kwargs)
+
+    def update(self, request, *args, **kwargs):
+        """
+        __Method__: PUT, PATCH
+
+        __URL__: $LINK:pushtarget-detail:instance_pk$
+
+        __Data__:
+
+        %(WRITABLE_SERIALIZER)s
+
+        __Response__:
+
+        %(SERIALIZER)s
+        """
+        return super(PushTargetViewSet, self).update(request, *args, **kwargs)
+
+    def destroy(self, *args, **kwargs):
```

Comment:
For push-targets, we'd better not add a delete function here, because if other resources use one push-target, you can't delete it. Create/update are enough.

Reply:
> if other resources use one push-target, you can't delete it

If you delete a push-target, it will be automatically removed from all fields that currently use it. Is this what you meant?

Reply:
No, I mean even when I try to delete a push-target, it can't be delete if this push-target is using in products/product-versions etc.
So I think we don't need this destroy function, maybe just the create and update are enough?

Reply:
Delete on a push-target now works fine, but it also removes itself from the allowed-push-targets (and the masked version).

@larry-shi What HTTP methods does need to be available for push-targets endpoint? Endpoint content-delivery-services only allows GET.

Reply:
@xiangge for me, @hluk 's solution is normal way.
@hluk at least get/create/update/delete, I am assuming we always should consider that way if no special reason?  what do you think?

Reply:
@hluk and @larry-shi, make sense to me.